### PR TITLE
Draft the statistical testing PLAN

### PR DESCRIPTION
**Pull request by nfrisby**

### CODE REVIEW

**Comment by nfrisby** [comment]
This is my summary of what we ran through during this morning's call. Please let me know if anything seems missing or out-of-place. Once we agree on this, I'll create some Issues for the first few steps.

File: ouroboros-consensus-test/PLAN-statistical-testing.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  - Praos inherently includes stochasticity. The leader schedule is determined
+    by a nonce. The test setup generators should effectively sample the initial
+    nonce from a uniform distribution.
+
+  - Chain propagation delay introduces stochasticity into any Ouroboros
+    protocol. If different candidate chains happen to arrive at a node at
+    essentially the same time, the node may have to select amongst
+    equally-preferable chains. The tie-breaker mechanism ought to be at least
+    partially stochastic. (It appears to be uniform from my limited empirical
+    analysis.)
```
Comment (mrBliss) [inline ouroboros-consensus-test/PLAN-statistical-testing.md:16]: Mock Praos just uses the `BlockNo` of the tips. For TPraos we look at the VRF hashes (it's more complicated). Do you think mock Praos needs a similar tie breaker?

File: ouroboros-consensus-test/PLAN-statistical-testing.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    relevant mock/test infrastructure (at least check for low-hanging fruit).
+
+  - Label CPs and truncate for CPs in the Mock Praos non-statistical
+    properties. (Maybe also Shelley, even before we have dedicated Shelley
+    statistical tests?)
+
+  - Introduce semi-synchrony, ie delta greater than 1. I'm disappointed that I
+    still don't have confidence regarding how exactly to inject these delays.
+    My only insight so far that feels particularly helpful is to only delay
+    RollForward messages and let all others traveral "instantly" (this is
```
Comment (mrBliss) [inline ouroboros-consensus-test/PLAN-statistical-testing.md:123]:
```suggestion
    RollForward messages and let all others travel "instantly" (this is
```

File: ouroboros-consensus-test/PLAN-statistical-testing.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    behind other messages, and -- similarly to priority inversion -- this may
+    delay a particular message beyond it's originally scheduled delay.
+
+## Miscellaneous Questions
+
+  - The above proposes focusing on the probability of a CP violation occuring
+    during some fixed number of slots. Should we instead focus on its arrival
+    rate? This would suggest choosing a "small" configuration that essentially
+    ensures a CP violation will occur within say the first 1000 slots --
+    otherwise the test may have extremely variable run-times.
```
Comment (mrBliss) [inline ouroboros-consensus-test/PLAN-statistical-testing.md:137]: I'm not sure what the consequences, can you explan?

File: ouroboros-consensus-test/PLAN-statistical-testing.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+      - The fraction of active slots should be f (regardless of stake
+        distribution, barring empty). Trivial to compute analytically.
+
+      - The leadership rate p_i of a node should match phi_f(its stake) (may
+        need hold the stake dist constant for this). Trivial to compute
+        analytically.
+
+      - The rate of multi-leader slots varies with f, n, and stake dist (for
+        small n) (constant stake dist again). Easy to compute analytically.
```
Comment (mrBliss) [inline ouroboros-consensus-test/PLAN-statistical-testing.md:149]: Sound good.

File: ouroboros-consensus-test/PLAN-statistical-testing.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+      - Nodes not leading should break any ties fairly when multiple of their
+        immediate neighbors are leading simultaneously. Should be uniform, I
+        think.
```
Comment (mrBliss) [inline ouroboros-consensus-test/PLAN-statistical-testing.md:153]: How would you do this? The BlockFetch client compares candidate chains to decide which block to download. How would you instrument this? Or would you do it at a higher level, i.e., looking at the produced and adopted blocks?

File: ouroboros-consensus-test/PLAN-statistical-testing.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+  - How to handle the "k+1st upstream header is one slot beyond stability
+    window" infelicity? How common is this? Easy to robustly detect? Etc.
```
Comment (mrBliss) [inline ouroboros-consensus-test/PLAN-statistical-testing.md:156]: I have no idea of how common it is (depends on `k` as you said). We should definitely label it.

File: ouroboros-consensus-test/PLAN-statistical-testing.md (hunk 1 above)
Comment (nfrisby) [reply ouroboros-consensus-test/PLAN-statistical-testing.md:16]: I think that's a question for the researchers. I do not recall seeing any tie breaking in the Ouroboros papers, so it currently seems preferable for the mock implementation to follow the papers as much as possible.

File: ouroboros-consensus-test/PLAN-statistical-testing.md (hunk 5 above)
Comment (nfrisby) [reply ouroboros-consensus-test/PLAN-statistical-testing.md:153]: Yep, at the higher-level. I can compare the `TraceForgedBlock`, `SwitchedFork` messages, and so on after the fact.

File: ouroboros-consensus-test/PLAN-statistical-testing.md (hunk 3 above)
Comment (nfrisby) [reply ouroboros-consensus-test/PLAN-statistical-testing.md:137]: I added a sentence here that explains that I'm currently not sure how to elaborate :)

File: ouroboros-consensus-test/PLAN-statistical-testing.md (hunk 2 above)
Comment (nfrisby) [reply ouroboros-consensus-test/PLAN-statistical-testing.md:123]: Thanks!

**Comment by nfrisby** [comment]
Closing. This has been superseded by the intervening ThreadNet rewrite work with Edsko. We took a different tack: we enforced Common Prefix by sacrificing Chain Growth. Thus, any wedge indicates a bug. We haven't focused on testing for Chain Growth; that'll perhaps need to be somewhat statistical again.

**Pull request closed without merging**